server/src/main/java/io/crate/execution/ddl/tables/TransportAddColumnAction.java

```diff
@@ -0,0 +1,182 @@
+            AcknowledgedResponse::new,
+            "add-column");
+        this.nodeContext = nodeContext;
+    }
+
+    @Override
+    public ClusterStateTaskExecutor<AddColumnRequest> clusterStateTaskExecutor(AddColumnRequest request) {
+        return new DDLClusterStateTaskExecutor<>() {
+            @Override
+            protected ClusterState execute(ClusterState currentState, AddColumnRequest request) throws Exception {
+                Metadata.Builder metadataBuilder = Metadata.builder(currentState.metadata());
+                if (request.isPartitioned()) {
+                    String templateName = PartitionName.templateName(request.relationName().schema(), request.relationName().name());
+                    IndexTemplateMetadata indexTemplateMetadata = currentState.metadata().templates().get(templateName);
+
+                    Map<String, Object> existingTemplateMeta = new HashMap<>();
+                    mergeDeltaIntoExistingMapping(existingTemplateMeta, request);
+
+                    IndexTemplateMetadata newIndexTemplateMetadata = DDLClusterStateHelpers.updateTemplate(
+                        indexTemplateMetadata,
+                        existingTemplateMeta,
+                        Collections.emptyMap(),
+                        Settings.EMPTY,
+                        (n, s) -> {},
+                        s -> true
+                    );
+                    metadataBuilder.put(newIndexTemplateMetadata);
+                }
+
+                currentState = updateMapping(currentState, request, metadataBuilder);
+                // ensure the new table can still be parsed into a DocTableInfo to avoid breaking the table.
+                new DocTableInfoFactory(nodeContext).create(request.relationName(), currentState);
+                return currentState;
+            }
+        };
+    }
+
+    /**
+     * Structure of the metadata is aligned with {@link AnalyzedTableElements#toMapping(AnalyzedTableElements)}
+     */
+    private ClusterState updateMapping(ClusterState currentState,
+                                       AddColumnRequest request,
+                                       Metadata.Builder metadataBuilder) throws IOException {
+        Index[] concreteIndices = resolveIndices(currentState, request.relationName().indexNameOrAlias());
+
+        for (Index index : concreteIndices) {
+            final IndexMetadata indexMetadata = currentState.metadata().getIndexSafe(index);
+            Map<String, Object> indexMapping = indexMetadata.mapping().sourceAsMap();
+
+            mergeDeltaIntoExistingMapping(indexMapping, request);
+
+            IndexMetadata.Builder imBuilder = IndexMetadata.builder(indexMetadata);
+            var updatedMetadata = new MappingMetadata(Constants.DEFAULT_MAPPING_TYPE, indexMapping); // throws IOException
+            imBuilder.putMapping(updatedMetadata).mappingVersion(1 + imBuilder.mappingVersion()); // Similar to MetadataMappingService
```

Comment:
I will remove this comment, just wanted to point out that version increment happens in 2 places - for IndexMetadata  (like in https://github.com/crate/crate/blob/master/server/src/main/java/org/elasticsearch/cluster/metadata/MetadataMappingService.java#L311) and Metadata itself